core/src/main/scala/zio/config/ReadModule.scala

```diff
@@ -234,22 +230,75 @@ private[config] trait ReadModule extends ConfigDescriptorModule {
+      // FIXME: Distinguish between re-use of structures vs recursive structures.
+      // If not distinguished, we miss out deep errors in the error report when config has same products/co-products repeated (than a recursion).
+      // However, this is not a major roadblock, as we keep deep error-reports in all other cases.
+
+      // FIXME: Remove following comments
+      //
+      // Why program-summary?: An early exit should be only for recursive values (`programSummary.contains(config)`)
+      // Such that if configuration isn't recursive, we report all errors at deeper levels.
+      //
+      // Example:
+      // Without a program summary, for a non-recursive structure, we get the following error structure:
+      //  {{{
+      //    MissingValue
+      //     path: a.a
+      //     Details: optional value
+      //  }}}
+      //
+      // With program summary, for a non-recursive structure, we get the following error structure
+      //
+      // {{{
+      //   MissingValue
+      //     path: a.a.b2
+      //     Details: optional value, value of type string
+      //   MissingValue
+      //     path: a.a.a2
+      //     Details: optional value, value of type string
+      // }}}
+      //
+      if (isEmptyConfigSource(config, keys.reverse) && programSummary.contains(config)) {
+        Left(ReadError.MissingValue(path.reverse, descriptions))
+      } else {
+        programSummary = programSummary ++ List(config)
+        config match {
+          case c @ Lazy(thunk) =>
+            loopAny(path, keys, thunk(), descriptions)
+          case c @ Default(_, _) =>
+            loopDefault(path, keys, c, descriptions)
+          case c @ Describe(_, message) =>
+            loopAny(path, keys, c.config, descriptions :+ message)
+          case c @ DynamicMap(_, _) => loopMap(path, keys, c, descriptions)
+          case c @ Nested(_, _, _) =>
+            loopNested(path, keys, c, descriptions)
+          case c @ Optional(config) =>
+            loopOptional(path, keys, c, descriptions)
+          case c @ OrElse(_, _) =>
+            loopOrElse(path, keys, c, descriptions)
+          case c @ OrElseEither(_, _) =>
+            loopOrElseEither(path, keys, c, descriptions)
+          case c @ Source(_, _) =>
+            loopSource(path, keys, c, descriptions)
+          case c @ Zip(_, _) =>
+            loopZip(path, keys, c, descriptions)
+          case c @ TransformOrFail(_, _, _) =>
+            loopXmapEither(path, keys, c, descriptions)
+          case c @ Sequence(_, _) =>
+            loopSequence(path, keys, c, descriptions)
+        }
       }
 
     loopAny(Nil, Nil, configuration, Nil).map(_.value)
+
+  }
+
+  private[config] def isEmptyConfigSource[A](
+    config: ConfigDescriptor[A],
+    keys: List[K]
+  ): Boolean = {
+    val sourceTrees = config.sources.map(_.getConfigValue(keys))
+    sourceTrees.forall(_ == PropertyTree.empty)
```

Comment:
PropertyTree.Empty is the only condition we need to verify and not `PropertyTree.isEmpty` because, `PropertyTree.isEmpty` returns true for empty `Sequence`